File: sdks/ts/packages/golem-ts-sdk/src/schema/agentStream.ts
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Golem Source License v1.1 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://license.golem.cloud/LICENSE
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

import { SchemaValueStream, type SchemaValueTree } from 'golem:core/types@2.0.0';
import {
  GuestSchemaValueStreamHandle,
  type GuestSchemaValueStream,
  schemaValueFromWit,
  schemaValueToWitAsync,
} from '../internal/schema-model';
import type { SchemaCodec } from './codec';

interface TypedStreamState<T> {
  readonly kind: 'typed';
  readonly source: AsyncIterable<T>;
  iterator?: AsyncIterator<T>;
  busy: boolean;
}

interface WireStreamState {
  readonly kind: 'wire';
  endpoint?: GuestSchemaValueStream;
  iterator?: AsyncIterator<SchemaValueTree>;
  readonly itemCodec: SchemaCodec;
  busy: boolean;
}

type AgentStreamState<T> = TypedStreamState<T> | WireStreamState;

const states = new WeakMap<object, AgentStreamState<unknown>>();

/**
 * A demand-driven, single-reader stream for values nested anywhere in an agent method input or
 * output.
 *
 * Reading is lazy and non-concurrent: wait for each operation before starting another. Normal
 * completion returns `{ done: true, value: undefined }`. Encoding or forwarding an `AgentStream`
 * transfers its ownership; the original object cannot be used again.
 *
 * For a stream received through a connected P3 agent invocation, calling `return()`, including the
 * call made by an early exit from `for await`, closes its readable endpoint. Calling `throw()` also
 * closes it and rejects with the same local reason; the reason is not transmitted to the producer.
 *
 * When a stream created with {@link AgentStream.from} is sent through a connected P3 invocation,
 * each accepted downstream write gates the next source pull. If a subsequent write observes that
 * the remote reader was dropped, the runtime stops pulling and invokes and awaits the source
 * iterator's `return()` exactly once. P3 does not interrupt an arbitrary pending source `next()` or
 * guarantee cleanup before a later agent invocation. A source or cleanup rejection fails the
 * active producer, write, or invocation session rather than appearing as clean EOF. P3 streams do
 * not carry a recoverable terminal error; use an item type such as `Result<T, E>` when errors must
 * be represented in the stream contract.
 */
export class AgentStream<T> implements AsyncIterable<T>, AsyncIterator<T> {
  private constructor() {}

  /**
   * Create a stream from a synchronous or asynchronous iterable without pulling it eagerly.
   *
   * The source iterator's `return()` is used for cleanup when either the local stream is closed or
   * a connected remote reader is dropped.
   */
  static from<T>(source: Iterable<T> | AsyncIterable<T>): AgentStream<T> {
    return createAgentStream({
      kind: 'typed',
      source: asAsyncIterable(source),
      busy: false,
    });
  }

  [Symbol.asyncIterator](): AsyncIterator<T> {
    return this;
  }

  /** Read the next item, or `{ done: true, value: undefined }` after clean completion. */
  async next(): Promise<IteratorResult<T>> {
    const state = streamState(this);
    if (state.busy) {
      throw new Error('an AgentStream operation is already in progress');
    }
    state.busy = true;
    try {
      if (state.kind === 'typed') {
        state.iterator ??= state.source[Symbol.asyncIterator]();
        const item = await state.iterator.next();
        return item.done ? { done: true, value: undefined } : item;
      }

      const iterator = await wireIterator(state);
      const item = await iterator.next();
      return item.done
        ? { done: true, value: undefined }
        : {
            done: false,
            value: state.itemCodec.fromValue(schemaValueFromWit(item.value)) as T,
          };
    } finally {
      state.busy = false;
    }
  }

  /**
   * Close the stream and its underlying iterator. `for await` invokes this automatically on an
   * early `break` or `return`.
   */
  async return(value?: unknown): Promise<IteratorResult<T>> {
    const state = streamState(this);
    if (state.busy) {
      throw new Error('an AgentStream operation is already in progress');
    }
    state.busy = true;
    try {
      states.delete(this);
      const iterator =
        state.kind === 'typed'
          ? (state.iterator ??= state.source[Symbol.asyncIterator]())
          : await wireIterator(state);
      if (iterator.return) {
        await iterator.return(value);
      }
      return { done: true, value: value as T };
    } finally {
      state.busy = false;
    }
  }

  /**
   * Fail local iteration and close a connected readable endpoint.
   *
   * Connected P3 streams reject with the same local reason, but do not transfer that reason to the
   * producer. Streams created with {@link AgentStream.from} delegate to the source iterator's
   * `throw()` when it provides one, but the `AgentStream` is consumed regardless of its result.
   */
  async throw(error?: unknown): Promise<IteratorResult<T>> {
    const state = streamState(this);
    if (state.busy) {
      throw new Error('an AgentStream operation is already in progress');
    }
    state.busy = true;
    try {
      states.delete(this);
      const iterator =
        state.kind === 'typed'
          ? (state.iterator ??= state.source[Symbol.asyncIterator]())
          : await wireIterator(state);
      if (iterator.throw) {
        return (await iterator.throw(error)) as IteratorResult<T>;
      }
      if (iterator.return) {
        await iterator.return();
      }
      throw error;
    } finally {
      state.busy = false;
    }
  }
}

/** @internal Move an AgentStream into a recursive schema value. */
export function agentStreamToHandle<T>(
  stream: AgentStream<T>,
  itemCodec: SchemaCodec,
): GuestSchemaValueStreamHandle {
  const state = streamState(stream);
  if (state.busy) {
    throw new Error('cannot transfer an AgentStream while an operation is in progress');
  }
  states.delete(stream);

  if (state.kind === 'wire') {
    if (state.endpoint !== undefined) {
      return new GuestSchemaValueStreamHandle(state.endpoint);
    }
    if (state.iterator !== undefined) {
      return new GuestSchemaValueStreamHandle({
        kind: 'native',
        value: iterableFromIterator(state.iterator),
      });
    }
    throw new Error('AgentStream was already transferred');
  }

  const source = state.iterator === undefined ? state.source : iterableFromIterator(state.iterator);
  return new GuestSchemaValueStreamHandle({
    kind: 'native',
    value: encodeItems(source, itemCodec),
  });
}

/** @internal Lift a recursive schema-value-stream handle into an AgentStream. */
export function agentStreamFromHandle<T>(
  handle: GuestSchemaValueStreamHandle,
  itemCodec: SchemaCodec,
): AgentStream<T> {
  const endpoint = handle.take();
  if (endpoint === undefined) {
    throw new Error('schema value stream was already transferred');
  }
  return createAgentStream({
    kind: 'wire',
    endpoint,
    itemCodec,
    busy: false,
  });
}

function createAgentStream<T>(state: AgentStreamState<T>): AgentStream<T> {
  const stream = Object.create(AgentStream.prototype) as AgentStream<T>;
  states.set(stream, state as AgentStreamState<unknown>);
  return stream;
}

function streamState<T>(stream: AgentStream<T>): AgentStreamState<T> {
  const state = states.get(stream) as AgentStreamState<T> | undefined;
  if (state === undefined) {
    throw new Error('AgentStream was already transferred or closed');
  }
  return state;
}

async function wireIterator(state: WireStreamState): Promise<AsyncIterator<SchemaValueTree>> {
  if (state.iterator !== undefined) {
    return state.iterator;
  }
  const endpoint = state.endpoint;
  if (endpoint === undefined) {
    throw new Error('AgentStream was already transferred');
  }
  state.endpoint = undefined;
  const source =
    endpoint.kind === 'native' ? endpoint.value : await SchemaValueStream.unwrap(endpoint.value);
  state.iterator = source[Symbol.asyncIterator]();
  return state.iterator;
}

function asAsyncIterable<T>(source: Iterable<T> | AsyncIterable<T>): AsyncIterable<T> {
  if (Symbol.asyncIterator in Object(source)) {
    return source as AsyncIterable<T>;
  }
  return {
    async *[Symbol.asyncIterator]() {
      yield* source as Iterable<T>;
    },
  };
}

function iterableFromIterator<T>(iterator: AsyncIterator<T>): AsyncIterable<T> {
  return {
    [Symbol.asyncIterator]: () => iterator,
  };
}

async function* encodeItems<T>(
  source: AsyncIterable<T>,
  itemCodec: SchemaCodec,
): AsyncIterable<SchemaValueTree> {
  for await (const item of source) {
    yield await schemaValueToWitAsync(itemCodec.toValue(item));
  }
}

File: sdks/go/golem/agentstream.go
// Copyright 2024-2026 Golem Cloud
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package golem

import (
	"errors"
	"fmt"
	"iter"
	"reflect"

	types "github.com/golemcloud/golem/sdks/go/golem/internal/wit/golem_core_types"
)

// Agent streams.
//
// An [AgentStream] carries a sequence of schema values between agents. It is
// the type a streaming method takes or returns:
//
//	var Tail = Logs.Method[TailIn, golem.AgentStream[LogLine]]("tail")
//
//	w, s := golem.NewAgentStream[LogLine]()
//	go func() {
//	    defer w.Close()              // clean completion; the reader sees EOF
//	    for _, line := range lines { // Write blocks: that is the backpressure
//	        if err := w.Write(line); err != nil { return }
//	    }
//	}()
//	return s
//
// Reading:
//
//	for line, err := range s.All() {
//	    if err != nil { ... }
//	}
//
// A stream endpoint is affine: handing it on transfers it rather than copying
// it, so a stream can be returned, or forwarded to another agent, but not both.
// Forwarding one you have not read costs nothing — the endpoint moves, no items
// are pumped through this component.
//
// Clean completion is end of input. A producer that fails must NOT close the
// stream cleanly, because the reader cannot tell that apart from success; model
// recoverable failures in the item type, as AgentStream[Result[T, E]].

var (
	// ErrStreamTransferred reports a stream used after it was handed on.
	ErrStreamTransferred = errors.New("golem: stream was already transferred")
	// ErrStreamPartiallyRead reports an attempt to forward a stream that has
	// already been read from. The items already taken cannot be put back.
	ErrStreamPartiallyRead = errors.New("golem: stream has already been read and cannot be forwarded")
	// ErrStreamClosed reports a write to a closed stream.
	ErrStreamClosed = errors.New("golem: stream is closed")
	// ErrReaderGone reports that the consumer dropped its end. A producer that
	// sees this should stop; it is cooperative, not an error in the data.
	ErrReaderGone = errors.New("golem: the stream's reader went away")
)

// treeSource is the reading half of a schema-value stream, and treeSink the
// writing half. They are structs of functions rather than interfaces on
// purpose: an interface call here builds an itab, and the linker then retains
// the method sets of every structurally-matching type — which off the wasm
// target drags host-call methods into a link that has no bodies for them. The
// indirection still keeps the chunking, EOF and ownership logic testable
// without a host.
type treeSource struct {
	read          func(dst []types.SchemaValueTree) uint32
	writerDropped func() bool
	drop          func()
	// reader is the host endpoint this source wraps, kept so an unread stream
	// can be handed straight back. It is nil for a source that did not come
	// from the host, which therefore cannot be transferred.
	reader any
}

type treeSink struct {
	write         func(items []types.SchemaValueTree) uint32
	readerDropped func() bool
	drop          func()
}

func (s treeSource) valid() bool { return s.read != nil }
func (s treeSink) valid() bool   { return s.write != nil }

// streamCodec converts between items and the value trees on the wire. It is
// derived from the SOURCE schema rather than from T alone, because a Go type
// does not determine its schema: []T is a list or a fixed list, string is a
// string or text or a path, int64 is an s64 or a quantity mantissa.
type streamCodec[T any] struct {
	encode func(T) (types.SchemaValueTree, error)
	decode func(types.SchemaValueTree) (T, error)
	// dispose releases endpoints an item still owns, for the case where the
	// item is rejected after it was built.
	dispose func(T)
}

// streamState is the shared state behind an [AgentStream]. Copies of the stream
// alias it, so a transfer made through one copy is seen by all of them — which
// is what makes the endpoint affine rather than merely conventionally so.
type streamState struct {
	src treeSource
	// pending holds the remainder of a read batch.
	pending []types.SchemaValueTree
	started bool
	taken   bool
	done    bool
	closed  bool
}

// AgentStream is a sequence of values streamed between agents.
type AgentStream[T any] struct {
	st    *streamState
	codec streamCodec[T]
}

// AgentStreamWriter produces the values of an [AgentStream].
type AgentStreamWriter[T any] struct {
	sink   treeSink
	codec  streamCodec[T]
	closed bool
}

// NewAgentStream creates a stream and its writer, with item conversion derived
// from T.
func NewAgentStream[T any]() (*AgentStreamWriter[T], AgentStream[T]) {
	return newAgentStreamWith(defaultStreamCodec[T]())
}

// NewAgentStreamWithCodec creates a stream whose items are converted by the
// supplied functions, so a caller holding the source schema can pin
// distinctions T alone does not carry. dispose may be nil.
func NewAgentStreamWithCodec[T any](
	encode func(T) (types.SchemaValueTree, error),
	decode func(types.SchemaValueTree) (T, error),
	dispose func(T),
) (*AgentStreamWriter[T], AgentStream[T]) {
	return newAgentStreamWith(streamCodec[T]{encode: encode, decode: decode, dispose: dispose})
}

func newAgentStreamWith[T any](c streamCodec[T]) (*AgentStreamWriter[T], AgentStream[T]) {
	sink, src := newStreamPair()
	return &AgentStreamWriter[T]{sink: sink, codec: c},
		AgentStream[T]{st: &streamState{src: src}, codec: c}
}

// StreamOf is a stream of known values, produced without a goroutine. It is
// mostly useful in tests and for small fixed results.
func StreamOf[T any](items ...T) AgentStream[T] {
	w, s := NewAgentStream[T]()
	go func() {
		defer func() { _ = w.Close() }()
		for _, item := range items {
			if err := w.Write(item); err != nil {
				return
			}
		}
	}()
	return s
}

// ProduceStream runs produce on its own goroutine and returns the stream it
// writes to. Returning nil closes the stream cleanly; returning an error
// abandons it, because a consumer must not read a failed production as a
// complete one.
func ProduceStream[T any](produce func(*AgentStreamWriter[T]) error) AgentStream[T] {
	w, s := NewAgentStream[T]()
	go func() {
		if err := produce(w); err != nil {
			// Abandon rather than close: dropping the writer without a clean
			// close is how the wire distinguishes an incomplete stream.
			w.abandon()
			return
		}
		_ = w.Close()
	}()
	return s
}

// Next reads the next value. It reports ok=false with a nil error at clean end
// of input; a non-nil error is a real failure and never means completion.
func (s AgentStream[T]) Next() (T, bool, error) {
	var zero T
	if s.st == nil {
		return zero, false, ErrStreamClosed
	}
	if s.st.taken {
		return zero, false, ErrStreamTransferred
	}
	if s.st.closed {
		return zero, false, ErrStreamClosed
	}
	s.st.started = true

	tree, ok, err := s.st.nextTree()
	if err != nil || !ok {
		return zero, false, err
	}
	v, err := s.codec.decode(tree)
	if err != nil {
		return zero, false, fmt.Errorf("golem: stream item: %w", err)
	}
	return v, true, nil
}

// nextTree pulls one value tree, refilling the batch when it runs out.
func (st *streamState) nextTree() (types.SchemaValueTree, bool, error) {
	for {
		if len(st.pending) > 0 {
			tree := st.pending[0]
			st.pending = st.pending[1:]
			return tree, true, nil
		}
		if st.done {
			return types.SchemaValueTree{}, false, nil
		}
		if !st.src.valid() {
			return types.SchemaValueTree{}, false, ErrStreamClosed
		}
		batch := make([]types.SchemaValueTree, 1)
		n := st.src.read(batch)
		if n == 0 {
			if st.src.writerDropped() {
				st.done = true
				return types.SchemaValueTree{}, false, nil
			}
			continue
		}
		st.pending = append(st.pending, batch[:n]...)
	}
}

// All iterates the stream. The loop ends at clean end of input; a failure is
// yielded once, with the zero value, and then the iteration stops.
func (s AgentStream[T]) All() iter.Seq2[T, error] {
	return func(yield func(T, error) bool) {
		for {
			v, ok, err := s.Next()
			if err != nil {
				var zero T
				yield(zero, err)
				return
			}
			if !ok {
				return
			}
			if !yield(v, nil) {
				return
			}
		}
	}
}

// Collect reads the stream to completion. Use it only where the sequence is
// known to be bounded — it defeats the point of streaming otherwise.
func (s AgentStream[T]) Collect() ([]T, error) {
	var out []T
	for v, err := range s.All() {
		if err != nil {
			return out, err
		}
		out = append(out, v)
	}
	return out, nil
}

// Close releases the reading endpoint. A producer observes it on its next
// write; it does not interrupt work already in flight.
func (s AgentStream[T]) Close() error {
	if s.st == nil || s.st.closed || s.st.taken {
		return nil
	}
	s.st.closed = true
	if s.st.src.valid() {
		s.st.src.drop()
	}
	return nil
}

// Write sends one value. It blocks until the consumer has room, which is what
// applies backpressure to the producer.
func (w *AgentStreamWriter[T]) Write(v T) error {
	if w.closed {
		return ErrStreamClosed
	}
	tree, err := w.codec.encode(v)
	if err != nil {
		if w.codec.dispose != nil {
			// The item was rejected after it was built, so anything it owns has
			// to be released here — nobody else has a reference to it.
			w.codec.dispose(v)
		}
		return fmt.Errorf("golem: stream item: %w", err)
	}
	batch := []types.SchemaValueTree{tree}
	for len(batch) > 0 {
		if w.sink.readerDropped() {
			return ErrReaderGone
		}
		n := w.sink.write(batch)
		batch = batch[n:]
	}
	return nil
}

// WriteAll sends several values in order, stopping at the first failure.
func (w *AgentStreamWriter[T]) WriteAll(vs ...T) error {
	for _, v := range vs {
		if err := w.Write(v); err != nil {
			return err
		}
	}
	return nil
}

// Close completes the stream. The reader sees end of input.
func (w *AgentStreamWriter[T]) Close() error {
	if w.closed {
		return nil
	}
	w.closed = true
	w.sink.drop()
	return nil
}

// abandon drops the writer without marking completion, which is how a failed
// production is distinguished from a finished one.
func (w *AgentStreamWriter[T]) abandon() {
	if w.closed {
		return
	}
	w.closed = true
	w.sink.drop()
}

// defaultStreamCodec converts items through the SDK's own codec for T.
func defaultStreamCodec[T any]() streamCodec[T] {
	c := defs.compile(reflect.TypeFor[T]())
	return streamCodec[T]{
		encode: func(v T) (types.SchemaValueTree, error) {
			if c.invalid != "" {
				return types.SchemaValueTree{}, errors.New(c.invalid)
			}
			return encodeWith(c, reflect.ValueOf(&v).Elem()), nil
		},
		decode: func(tree types.SchemaValueTree) (T, error) {
			var out T
			if c.invalid != "" {
				return out, errors.New(c.invalid)
			}
			d := decoder{nodes: tree.ValueNodes}
			slot := reflect.ValueOf(&out).Elem()
			if err := c.decode(&d, slot, tree.Root); err != nil {
				return out, err
			}
			return out, nil
		},
	}
}
